src/numerics/static_condensation.C

```diff
@@ -0,0 +1,455 @@
+          const Elem & elem,
+          const unsigned int node_num,
+          std::vector<dof_id_type> & dof_indices,
+          const dof_id_type field_dof)
+  {
+    computeElemInteriorDofsField(elem_interior_dofs)(elem, node_num, dof_indices, field_dof);
+
+    if (node_num != invalid_uint && !elem.is_internal(node_num))
+    {
+      elem_trace_dofs.push_back(field_dof);
+      const auto & nd_ref = elem.node_ref(node_num);
+      if (nd_ref.processor_id() == _dof_map.processor_id())
+        local_trace_dofs.insert(field_dof);
+      else
+        nonlocal_trace_dofs[nd_ref.processor_id()].insert(field_dof);
+    }
+  };
+
+  for (auto elem : _mesh.active_local_element_ptr_range())
+  {
+    elem_interior_dofs.clear();
+    elem_trace_dofs.clear();
+    unsigned int boundary_dof_size = 0;
+    unsigned int interior_dof_size = 0;
+    auto & local_data = _elem_to_local_data[elem->id()];
+
+    const auto sub_id = elem->subdomain_id();
+    for (const auto vg : make_range(_dof_map.n_variable_groups()))
+    {
+      const auto & var_group = _dof_map.variable_group(vg);
+      if (!var_group.active_on_subdomain(sub_id))
+        continue;
+
+      for (const auto v : make_range(var_group.n_variables()))
+      {
+        auto & var_data = local_data.var_to_data[v];
+        _dof_map.dof_indices(
+            elem, elem_dofs, v, scalar_dofs_functor, field_dofs_functor, elem->p_level());
+        var_data.boundary_dofs_offset = boundary_dof_size;
+        var_data.interior_dofs_offset = interior_dof_size;
+        var_data.num_boundary_dofs = elem_trace_dofs.size();
+        var_data.num_interior_dofs = elem_interior_dofs.size();
+        boundary_dof_size += var_data.num_boundary_dofs;
+        interior_dof_size += var_data.num_interior_dofs;
+        local_data.reduced_space_indices = elem_trace_dofs;
+      }
+    }
+
+    local_data.Aii.resize(interior_dof_size, interior_dof_size);
+    local_data.Aib.resize(interior_dof_size, boundary_dof_size);
+    local_data.Abi.resize(boundary_dof_size, interior_dof_size);
+    local_data.Abb.resize(boundary_dof_size, boundary_dof_size);
+  }
+  _local_trace_dofs.assign(local_trace_dofs.begin(), local_trace_dofs.end());
+  local_trace_dofs.clear();
+
+  //
+  // Build the reduced system data
+  //
+
+  const dof_id_type n_local = _local_trace_dofs.size();
+  dof_id_type n = n_local;
+  _dof_map.comm().sum(n);
+  _reduced_sys_mat = SparseMatrix<Number>::build(_dof_map.comm());
+  auto sp =
+      _dof_map.build_sparsity(_mesh, /*calculate_constrained=*/false, /*trace_dofs_only=*/true);
+  const auto & nnz = sp->get_n_nz();
+  const auto & noz = sp->get_n_oz();
+  const auto nz = nnz.empty() ? dof_id_type(0) : *std::max_element(nnz.begin(), nnz.end());
+  const auto oz = noz.empty() ? dof_id_type(0) : *std::max_element(noz.begin(), noz.end());
+  _reduced_sys_mat->init(n, n, n_local, n_local, nz, oz);
+  _reduced_solver = LinearSolver<Number>::build(_dof_map.comm());
```

Comment:
Is this limited to linear problems for now?  Any idea how hard it would be to extend?

Reply:
The only problems I work on are nonlinear 😄 So yea per https://github.com/libMesh/libmesh/pull/3883#discussion_r1657889786 and the list in #3891 this will be useful in a nonlinear context shortly after I wrap this class in a preconditioner